### ISSUE TITLE
PKCE support for OAuth 2.0

### ISSUE THREAD

**Issue by tacovandenbroek** [a85/POSTMan-Chrome-Extension#3825]
## App Details:
```
Postman for Linux
Version 5.3.2
linux 4.10.0-38-generic / x64
```

## Issue Report:
I'ld like to be able to use the "Proof Key for Code Exchange" extension of the OAuth 2.0 spec for Authorization Code Grant flows.

RFC: https://tools.ietf.org/html/rfc7636

**Comment by prashantagarwal** [comment]
@tacovandenbroek Thanks for requesting this feature.

**Comment by Amzani** [comment]
Hello guys,
Any news about this feature ?

**Comment by jon-whit** [comment]
Looking for this feature as well! Any updates?

**Comment by Hamitamaru** [comment]
Any update on this?

**Comment by chrisketelaar** [comment]
👍 from me, gonna need this feature in the near future.

**Comment by Laffarli** [comment]
:+1: Thanks in advance for adding PKCE support.

**Comment by JacobPoldenValtech** [comment]
Any update on this?

**Comment by JacobPoldenValtech** [comment]
What's going on with this. This is part of the RFC spec. Issues getting fixed include things like fonts and colours when this kind of missing feature is still outstanding? The prioritisation seems off.

**Comment by WolfspiritM** [comment]
We need this aswell...any news?

**Comment by davidhjones** [comment]
We'll need this as well. @shamasis  / @numaanashraf  any update on plans for PKCE support?

**Comment by ralphbragg-ob** [comment]
Hi - PKCE support is highly recommded by the Financial Grade Working group of the Open ID Foundation. More and more industries will require PKCE from clients so again +1

**Comment by brendon-stephens** [comment]
This would be a useful feature  + 1

**Comment by SamuelTissot** [comment]
+1 this would be great.

**Comment by thomas-lucas1985** [comment]
Would love to see this feature as well

**Comment by WellspringCS** [comment]
[Authorization Code flow is now the recommended flow for SPAs.](https://tools.ietf.org/html/draft-ietf-oauth-security-topics-11)

Note this quote from that link: "Clients utilizing the authorization grant type MUST use PKCE..."

Having switched to this recommended flow from Implicit, I now find that Postman cannot acquire a token anymore. :-(

Put my name in the hat of those waiting for this feature. Thanks!

**Comment by danni-popova** [comment]
+1

**Comment by dlorych** [comment]
+1

**Comment by HenriLangenhoven** [comment]
+1

**Comment by peval27** [comment]
+1

**Comment by richardkarlsen** [comment]
Ping 😄

**Comment by OwenVey** [comment]
+1

**Comment by tuandao9** [comment]
+1

**Comment by clementcur** [comment]
While waiting for the official implementation in Postman, I've developped a quick&dirty workaround with pre-request scripts and tests. Here is an example with **Keycloak** as the OPID Provider. Tokens content is outputed to the Postman Console (at request 3).

- collection variables:
  - AUTHENTICATION_URL: https://domain.com/auth/realms/myrealm/protocol/openid-connect/auth
  - TOKEN_URL: https://domain.com/auth/realms/myrealm/protocol/openid-connect/token
  - CLIENT_ID: my_client_id
  - REDIRECT_URI: http://localhost:3000
  - USERNAME: myusername
  - PASSWORD: myp@ssw0rd
  - READONLY_AUTHORIZATION_URL: _leave empty_
  - READONLY_CODE_VERIFIER: _leave empty_
  - READONLY_CODE_CHALLENGE: _leave empty_
  - READONLY_AUTHORIZATION_CODE: _leave empty_
  - READONLY_ACCESS_TOKEN: _leave empty_
  - READONLY_ID_TOKEN: _leave empty_

- Request 1:
  - GET -> `{{AUTHENTICATION_URL}}`
  - params 
    - client_id: `{{CLIENT_ID}}`
    - redirect_uri: `{{REDIRECT_URI}}`
    - response_type: code
    - scope: openid
    - code_challenge: `{{CODE_CHALLENGE}}`
    - code_challenge_method: S256
  - pre-request script
    ```
    function generateCodeVerifier() {
        return generateRandomString(96);
    }
    
    function generateRandomString(length) {
        var text = "";
        var possible = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";
        for (var i = 0; i < length; i++) {
            text += possible.charAt(Math.floor(Math.random() * possible.length));
        }
        return text;
    }

    function generateCodeChallenge(code_verifier) {
        return CryptoJS.SHA256(code_verifier);
    }

    function base64URL(string) {
        return string.toString(CryptoJS.enc.Base64).replace(/=/g, '').replace(/\+/g, '-').replace(/\//g, '_')
    }

    var verifier = base64URL(generateCodeVerifier());
    var challenge = base64URL(generateCodeChallenge(verifier));

    // clearing all existing cookies
    const cookieJar = pm.cookies.jar();
    cookieJar.clear(pm.collectionVariables.get("AUTHENTICATION_URL"));
    cookieJar.clear(pm.collectionVariables.get("TOKEN_URL"));

    pm.collectionVariables.set("READONLY_CODE_VERIFIER", verifier);
    pm.collectionVariables.set("READONLY_CODE_CHALLENGE", challenge);
    pm.collectionVariables.unset("READONLY_AUTHORIZATION_URL");
    pm.collectionVariables.unset("READONLY_AUTHORIZATION_CODE");
    pm.collectionVariables.unset("READONLY_ACCESS_TOKEN");
    pm.collectionVariables.unset("READONLY_ID_TOKEN");
    ```
  - tests
    ```
    var response = pm.response.text();

    var pattern = /action=".*?"/g;

    var found = pattern.exec(response);
    var url = "";

    if (!!found && found.length > 0) {
      url = found[0].substring(8, found[0].length-1);
      url = url.replace(/&amp;/g, '&');
      pm.collectionVariables.set("READONLY_AUTHORIZATION_URL", url);
  
      pm.test("All good! => Go to step 2.", function () {
        pm.expect(url).to.not.be.undefined;
        pm.expect(url).to.not.equal("");
      });
    } else {
      pm.expect.fail("Unable to extract Authorization URL from HTML, check the collection variables.");
    }
    ```

- Request 2:
  - GET -> `{{READONLY_AUTHORIZATION_URL}}`
  - headers
    - content-type: application/x-www-form-urlencoded
  - body
    - username: `{{USERNAME}}`
    - password: `{{PASSWORD}}`
  - tests
    ```
    var redirect = postman.getResponseHeader("Location");

    if (!!redirect && redirect.length > 0) {
      var redirectSplit = redirect.split("code=");
  
      if (!!redirectSplit && redirectSplit.length >= 2) {
          var authorizationCode = redirectSplit[1];
          pm.collectionVariables.set("READONLY_AUTHORIZATION_CODE", authorizationCode);
    
          pm.test("All good! => Go to step 3.", function () {
            pm.expect(authorizationCode).to.not.be.undefined;
          });
      } else {
        pm.expect.fail("Unable to login, either your username/password are wrong, or you are already connected.");
      }
    } else {
      pm.expect.fail("Unable to login, either your username/password are wrong, or you are already connected.");
    }
    ```
  - settings
    - uncheck "Automatically follow redirects"

- Request 3:
  - POST -> `{{TOKEN_URL}}`
  - headers
    - content-type: application/x-www-form-urlencoded
  - body
    - code: `{{READONLY_AUTHORIZATION_CODE}}`
    - grant_type: authorization_code
    - client_id: `{{CLIENT_ID}}`
    - redirect_uri: `{{REDIRECT_URI}}`
    - code_verifier: `{{READONLY_CODE_VERIFIER}}`
  - tests
    ```
    // Util method to decode JWTs
    function jwt_decode(jwt) {
        var parts = jwt.split('.'); // header, payload, signature
        return JSON.parse(atob(parts[1])); // only parsing payload
    }

    var jsonData = pm.response.json();
    var jsonAT = jsonData.hasOwnProperty('access_token')
      ? jwt_decode(jsonData.access_token)
      : undefined;
    var jsonRT = jsonData.hasOwnProperty('refresh_token')
      ? jwt_decode(jsonData.refresh_token)
      : undefined;
    var jsonIDT = jsonData.hasOwnProperty('id_token')
      ? jwt_decode(jsonData.id_token)
      : undefined;

    if (jsonData && !!jsonData.error) {
      pm.expect.fail("Error" + !!jsonData.error_description ? jsonData.error_description : "");
    } else {
      pm.collectionVariables.set("READONLY_ACCESS_TOKEN", jsonData.access_token);
      pm.collectionVariables.set("READONLY_ID_TOKEN", jsonData.id_token);

      pm.test("Access Token obtained, you can now securely call the API of your choice!", function () {
        console.log("Access Token:");
        console.log(jsonAT);
        pm.expect(jsonAT).to.not.be.undefined;
      });

      pm.test("Refresh Token obtained, you can now refresh your Access Token!", function () {
        console.log("Refresh Token:");
        console.log(jsonRT);
        pm.expect(jsonRT).to.not.be.undefined;
      });

      pm.test("ID Token obtained, you can now be identified when calling the API of your choice!", function () {
        console.log("ID Token:");
        console.log(jsonIDT);
        pm.expect(jsonIDT).to.not.be.undefined;
      });
    }
    ```

- Request 4:
  - GET -> whatever
  - authorization
    - type: OAuth 2.0
    - Add to: Request Headers
    - Access Token: `{{READONLY_ACCESS_TOKEN}}`

**Comment by ilya-chumakov** [comment]
Still waiting for the official implementation :)

**Comment by jwvance** [comment]
My team needs this feature! Any updates?

**Comment by PhilipSkinner** [comment]
We're having to deploy an auth code flow client thats just for testers into our environment(s) due to this lack of support.

+1 on this feature, should be relatively simple to implement.

**Comment by illialarka** [comment]
That's not okay!
Authorization Code with PKCE is a RFC PROPOSED STANDARD.
Honestly, most of modern web applications use this flow.

**Comment by murbanowicz** [comment]
Finally we got dev assigned. @vvs11 - can you give us any update on this please?

**Comment by vikiCoder** [comment]
Hello everyone,

We are excited to announce that PKCE support for OAuth 2.0 is added in the latest version of [Postman Canary (v7.23.0-canary01)](https://www.postman.com/downloads/canary). To use it, open the `Get New Access Token` dialogue and select `Authorization Code (With PKCE)` as `Grant type`. You can select the `Code Challenge Method` to be either `Plain` or `SHA-256` and also optionally provide the custom `Code Verifier`. Download the latest Postman Canary app to check it out. Any feedbacks are welcome 🙂

![image](https://user-images.githubusercontent.com/16588032/79312355-65aefc80-7f1c-11ea-8a79-33fd0d8a6dbc.png)

**Comment by donbowman** [comment]
Thanks for this.
Tried it, infortunately the built-in browser is not sufficient, returning:
```
This browser or app may not be secure. Learn more

Try using a different browser. If you’re already using a supported browser, you can refresh your screen and try again to sign in.
```
![image](https://user-images.githubusercontent.com/5131923/79340347-2a9ad200-7ef8-11ea-94d4-9580402f1739.png)

A *much more common* technique is to open a local port (in postman, e.g. port 8888), and then cal xdg-open on the real browser, w/ the redirect being the 8888. This means the real browser (which is likely already signed in) will finish the flow back to postman.

this is how e.g. gcloud works.

**Comment by vikiCoder** [comment]
@donbowman You are right, it is a [known issue](https://github.com/postmanlabs/postman-app-support/issues/8059). We are working on allowing token generation through the browser. It will be released soon. Follow [this](https://github.com/postmanlabs/postman-app-support/issues/8059) thread for updates. Here's a glimpse of that feature:

![image](https://user-images.githubusercontent.com/16588032/79343331-e5a08b00-7f4b-11ea-94cb-d79eef74cbe1.png)

**Comment by RickC37** [comment]
In my system it worked perfectly. Thank you @vikiCoder 🙏

**Comment by donbowman** [comment]
i also wish it would use the openid discovery document. /.well-known/openid-configuration.  that would remove the need to fill in many fields.
its also confusing to call the redirect-uri 'callback url' since that has a (different) meaning.

consider suggest `urn:ietf:wg:oauth:2.0:oob` as the redirect uri, this is from the standard for this type of app when doing the out-of-band i think.
(doc here https://developers.google.com/identity/protocols/oauth2/native-app)

**Comment by vikiCoder** [comment]
@donbowman Thanks for the suggestion. Feature request for OpenID discovery is tracked [here](https://github.com/postmanlabs/postman-app-support/issues/2361). However we haven't planned it yet. Stay tuned for updates🙂

**Comment by vikiCoder** [comment]
PKCE support for OAuth 2.0 is now also available in the latest version of [Postman app (v7.23.0)](https://www.postman.com/downloads/). See [this comment](https://github.com/postmanlabs/postman-app-support/issues/3825#issuecomment-613880232) for the details of this feature. Please update to the latest version of Postman app to try it out.

**Comment by hellowoodoo** [comment]
> While waiting for the official implementation in Postman, I've developped a quick&dirty workaround with pre-request scripts and tests. Here is an example with **Keycloak** as the OPID Provider. Tokens content is outputed to the Postman Console (at request 3).
> 
> * collection variables:
>   
>   * AUTHENTICATION_URL: https://domain.com/auth/realms/myrealm/protocol/openid-connect/auth
>   * TOKEN_URL: https://domain.com/auth/realms/myrealm/protocol/openid-connect/token
>   * CLIENT_ID: my_client_id
>   * REDIRECT_URI: http://localhost:3000
>   * USERNAME: myusername
>   * PASSWORD: myp@ssw0rd
>   * READONLY_AUTHORIZATION_URL: _leave empty_
>   * READONLY_CODE_VERIFIER: _leave empty_
>   * READONLY_CODE_CHALLENGE: _leave empty_
>   * READONLY_AUTHORIZATION_CODE: _leave empty_
>   * READONLY_ACCESS_TOKEN: _leave empty_
>   * READONLY_ID_TOKEN: _leave empty_
> * Request 1:
>   
>   * GET -> `{{AUTHENTICATION_URL}}`
>   * params
>     
>     * client_id: `{{CLIENT_ID}}`
>     * redirect_uri: `{{REDIRECT_URI}}`
>     * response_type: code
>     * scope: openid
>     * code_challenge: `{{CODE_CHALLENGE}}`
>     * code_challenge_method: S256
>   * pre-request script
>     ```
>     function generateCodeVerifier() {
>         return generateRandomString(96);
>     }
>     
>     function generateRandomString(length) {
>         var text = "";
>         var possible = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";
>         for (var i = 0; i < length; i++) {
>             text += possible.charAt(Math.floor(Math.random() * possible.length));
>         }
>         return text;
>     }
>     
>     function generateCodeChallenge(code_verifier) {
>         return CryptoJS.SHA256(code_verifier);
>     }
>     
>     function base64URL(string) {
>         return string.toString(CryptoJS.enc.Base64).replace(/=/g, '').replace(/\+/g, '-').replace(/\//g, '_')
>     }
>     
>     var verifier = base64URL(generateCodeVerifier());
>     var challenge = base64URL(generateCodeChallenge(verifier));
>     
>     // clearing all existing cookies
>     const cookieJar = pm.cookies.jar();
>     cookieJar.clear(pm.collectionVariables.get("AUTHENTICATION_URL"));
>     cookieJar.clear(pm.collectionVariables.get("TOKEN_URL"));
>     
>     pm.collectionVariables.set("READONLY_CODE_VERIFIER", verifier);
>     pm.collectionVariables.set("READONLY_CODE_CHALLENGE", challenge);
>     pm.collectionVariables.unset("READONLY_AUTHORIZATION_URL");
>     pm.collectionVariables.unset("READONLY_AUTHORIZATION_CODE");
>     pm.collectionVariables.unset("READONLY_ACCESS_TOKEN");
>     pm.collectionVariables.unset("READONLY_ID_TOKEN");
>     ```
>   * tests
>     ```
>     var response = pm.response.text();
>     
>     var pattern = /action=".*?"/g;
>     
>     var found = pattern.exec(response);
>     var url = "";
>     
>     if (!!found && found.length > 0) {
>       url = found[0].substring(8, found[0].length-1);
>       url = url.replace(/&amp;/g, '&');
>       pm.collectionVariables.set("READONLY_AUTHORIZATION_URL", url);
>     
>       pm.test("All good! => Go to step 2.", function () {
>         pm.expect(url).to.not.be.undefined;
>         pm.expect(url).to.not.equal("");
>       });
>     } else {
>       pm.expect.fail("Unable to extract Authorization URL from HTML, check the collection variables.");
>     }
>     ```
> * Request 2:
>   
>   * GET -> `{{READONLY_AUTHORIZATION_URL}}`
>   * headers
>     
>     * content-type: application/x-www-form-urlencoded
>   * body
>     
>     * username: `{{USERNAME}}`
>     * password: `{{PASSWORD}}`
>   * tests
>     ```
>     var redirect = postman.getResponseHeader("Location");
>     
>     if (!!redirect && redirect.length > 0) {
>       var redirectSplit = redirect.split("code=");
>     
>       if (!!redirectSplit && redirectSplit.length >= 2) {
>           var authorizationCode = redirectSplit[1];
>           pm.collectionVariables.set("READONLY_AUTHORIZATION_CODE", authorizationCode);
>     
>           pm.test("All good! => Go to step 3.", function () {
>             pm.expect(authorizationCode).to.not.be.undefined;
>           });
>       } else {
>         pm.expect.fail("Unable to login, either your username/password are wrong, or you are already connected.");
>       }
>     } else {
>       pm.expect.fail("Unable to login, either your username/password are wrong, or you are already connected.");
>     }
>     ```
>   * settings
>     
>     * uncheck "Automatically follow redirects"
> * Request 3:
>   
>   * POST -> `{{TOKEN_URL}}`
>   * headers
>     
>     * content-type: application/x-www-form-urlencoded
>   * body
>     
>     * code: `{{READONLY_AUTHORIZATION_CODE}}`
>     * grant_type: authorization_code
>     * client_id: `{{CLIENT_ID}}`
>     * redirect_uri: `{{REDIRECT_URI}}`
>     * code_verifier: `{{READONLY_CODE_VERIFIER}}`
>   * tests
>     ```
>     // Util method to decode JWTs
>     function jwt_decode(jwt) {
>         var parts = jwt.split('.'); // header, payload, signature
>         return JSON.parse(atob(parts[1])); // only parsing payload
>     }
>     
>     var jsonData = pm.response.json();
>     var jsonAT = jsonData.hasOwnProperty('access_token')
>       ? jwt_decode(jsonData.access_token)
>       : undefined;
>     var jsonRT = jsonData.hasOwnProperty('refresh_token')
>       ? jwt_decode(jsonData.refresh_token)
>       : undefined;
>     var jsonIDT = jsonData.hasOwnProperty('id_token')
>       ? jwt_decode(jsonData.id_token)
>       : undefined;
>     
>     if (jsonData && !!jsonData.error) {
>       pm.expect.fail("Error" + !!jsonData.error_description ? jsonData.error_description : "");
>     } else {
>       pm.collectionVariables.set("READONLY_ACCESS_TOKEN", jsonData.access_token);
>       pm.collectionVariables.set("READONLY_ID_TOKEN", jsonData.id_token);
>     
>       pm.test("Access Token obtained, you can now securely call the API of your choice!", function () {
>         console.log("Access Token:");
>         console.log(jsonAT);
>         pm.expect(jsonAT).to.not.be.undefined;
>       });
>     
>       pm.test("Refresh Token obtained, you can now refresh your Access Token!", function () {
>         console.log("Refresh Token:");
>         console.log(jsonRT);
>         pm.expect(jsonRT).to.not.be.undefined;
>       });
>     
>       pm.test("ID Token obtained, you can now be identified when calling the API of your choice!", function () {
>         console.log("ID Token:");
>         console.log(jsonIDT);
>         pm.expect(jsonIDT).to.not.be.undefined;
>       });
>     }
>     ```
> * Request 4:
>   
>   * GET -> whatever
>   * authorization
>     
>     * type: OAuth 2.0
>     * Add to: Request Headers
>     * Access Token: `{{READONLY_ACCESS_TOKEN}}`

Dear clemetcur,

I tried to run your script, but the CODE_CHALLENGE is not defined during the run. 
Can you help me, please ho to resolve the problem?
